app/src/operations_endpoints.h

```diff
@@ -0,0 +1,402 @@
+   * submit asynchronous operations' results.
+   *
+   * We use the node table to determine what port the service is listening on.
+   * This works even if cchost was given port 0 in its configuration, as it
+   * will have updated the node information after the port is bound.
+   */
+  std::string get_bind_address(
+    ccfapp::AbstractNodeContext& context, kv::ReadOnlyTx& tx)
+  {
+    auto nodes = tx.ro<ccf::Nodes>(ccf::Tables::NODES);
+    ccf::NodeId node_id = context.get_node_id();
+    std::optional<ccf::NodeInfo> node_info = nodes->get(node_id);
+
+    if (node_info.has_value() && !node_info->rpc_interfaces.empty())
+    {
+      // cchost can listen on multiple interfaces. This arbitrarily uses the
+      // first one, in alphabetical order.
+      const auto& primary_interface = node_info->rpc_interfaces.begin()->second;
+
+      // Note that bind_address can be 0.0.0.0. This is fine here
+      // as Linux routes that address to localhost.
+      return primary_interface.bind_address;
+    }
+    else
+    {
+      throw std::runtime_error("Could not callback URL");
+    }
+  }
+
+  /**
+   * Mark the start of an asynchronous operation.
+   *
+   * Asynchronous operations are useful for tasks which depend on an external
+   * process to execute. This transaction will record the start of the operation
+   * and set its status to "pending". When the external process completes, it
+   * must invoke a callback URL, at which point more processing will happen
+   * inside the ledger.
+   *
+   * Ideally we'd kick-off the background process here, but we need access
+   * to the operation ID to construct the callback URL, and we don't know it
+   * yet. We use a local commit handler (operation_locally_committed_func)
+   * to run code after the transaction has been committed. The trigger argument
+   * is saved and will be called from the commit handler.
+   *
+   * The callback_context object may be used to preserve information about the
+   * asynchronous operation, and will be available to the handler when the
+   * background process completes and invokes the callback URL.
+   */
+  void start_asynchronous_operation(
+    ccfapp::AbstractNodeContext& node_context,
+    ccf::endpoints::EndpointContext& endpoint_context,
+    nlohmann::json callback_context,
+    TriggerAsynchronousOperation trigger)
+  {
+    auto table =
+      endpoint_context.tx.template rw<OperationsTable>(OPERATIONS_TABLE);
+    table->put(OperationLog{
+      .status = OperationStatus::Pending,
+      .context = std::move(callback_context),
+    });
+
+    // The AppData allows us to propagate data to the handler. We use it to keep
+    // a reference to the trigger lambda and the bind address. The latter is
+    // needed to build the callback URL, but needs read access to the KV, which
+    // we won't have later on in the commit handler.
+    get_app_data(endpoint_context.rpc_ctx).asynchronous_operation =
+      AsynchronousOperation{
+        .trigger = trigger,
```

Comment:
This is definitely not something we'd be able to do with CCF's external executors, at most they would allow some JSON type user data.

Reply:
Yeah that's probably right.

I think I'd rather keep it as is for now. Using a lambda here is nice, since it means the operations code can be very agnostic of how they are actually implemented, and doesn't need to explicitly capture references to CCF state (eg. the AbstractNodeContext needed to get the AbstractHostProcesses object).

Reply:
Based on `https://github.com/microsoft/CCF/issues/4996`, this seems fine actually.

If anything, it looks like we won't even need the AppData anymore and the two phases of the request will happen in the same context.